libcache/cache.c

```diff
@@ -0,0 +1,749 @@
+{
+	unsigned char flags = 0;
+	ssize_t read = 0, readPos = 0, written = 0;
+	void *temp = NULL;
+	uint64_t tempAddr = 0;
+	size_t left = cache->lineSize;
+	cacheline_t line, *linePtr = NULL;
+
+	temp = malloc(cache->lineSize * sizeof(unsigned char));
+
+	if (temp == NULL) {
+		return -1;
+	}
+
+	tempAddr = addr;
+
+	while (left > 0) {
+		read = cache->readCb(tempAddr, (unsigned char *)temp + readPos, left);
+
+		if (read == -1) {
+			return -1;
+		}
+
+		left -= read;
+		tempAddr += read;
+		readPos += read;
+	}
+
+	SET_VALID(flags);
+	line.tag = tag;
+	line.addr = addr;
+	line.data = temp;
+	line.flags = flags;
+	line.offset = offset;
+
+	written = cache_writeToSet(cache->writeCb, &cache->sets[index], &line, addr, cache->lineSize, cache->srcMemSize);
+
+	return written;
+}
+
+
+ssize_t cache_write(cachectx_t *cache, uint64_t addr, void *buffer, size_t count, int policy)
+{
+	int ret = -1;
+	ssize_t position = 0, readPos = 0, written = 0;
+	void *data = NULL, *dest = NULL;
+	unsigned char flags = 0;
+	uint64_t index = 0, offset = 0, tag = 0, tempAddr = 0;
+	size_t pieceSize = 0, left = 0, remainder = 0, leftToRead = 0;
+	cacheline_t line, *linePtr = NULL;
+
+	if (cache == NULL || buffer == NULL || (policy != LIBCACHE_WRITE_BACK && policy != LIBCACHE_WRITE_THROUGH)) {
+		return -1;
+	}
+
+	if (count == 0) {
+		return 0;
+	}
+
+	if (addr > cache->srcMemSize) {
+		return 0;
+	}
+	else if ((addr < cache->srcMemSize) && (addr + count > cache->srcMemSize)) {
+		count = cache->srcMemSize - addr;
+	}
+
+	left = count;
+
+	leftToRead = cache->lineSize;
+	offset = cache_compOffset(cache, addr);
+	remainder = (left - (cache->lineSize - offset)) % cache->lineSize;
+
+	mutexLock(cache->lock);
+
+	while (left > 0) {
+		index = cache_compSet(cache, addr);
+		tag = cache_compTag(cache, addr);
+
+		pieceSize = cache_compPieceSize(left, cache->lineSize, &addr, &offset, count, remainder);
+
+		linePtr = cache_findLine(&cache->sets[index], tag, LIBCACHE_TIMESTAMPS_UPDATE);
+
+		/* cache miss */
+		if (linePtr == NULL) {
+			if (pieceSize != cache->lineSize) {
+				ret = cache_readFromMemOnMiss(cache, addr, index, tag, offset);
+				printf("ret: %d\n", ret);
+				if (ret != cache->lineSize) {
+					if (ret != -1) {
+						position += ret;
+					}
+					break;
+				}
+
+				linePtr = cache_findLine(&cache->sets[index], tag, LIBCACHE_TIMESTAMPS_NO_UPDATE);
+				memcpy((unsigned char *)linePtr->data + offset, (unsigned char *)buffer + position, pieceSize);
+			}
+			else {
+				data = calloc(cache->lineSize, sizeof(unsigned char));
+
+				if (data == NULL) {
+					break;
+				}
+
+				memcpy((unsigned char *)data + offset, (unsigned char *)buffer + position, pieceSize);
+
+				SET_VALID(flags);
+				line.tag = tag;
+				line.addr = addr;
+				line.data = data;
+				line.flags = flags;
+				line.offset = offset;
+
+				written = cache_writeToSet(cache->writeCb, &cache->sets[index], &line, addr, cache->lineSize, cache->srcMemSize);
```

Comment:
This stuff seems to be always closely related to cache_writeToSet. How about making this function allocate new cache line instead, so all operation are there instead (and change name to e.g. cache_allocateLine)? This would make memory management less prone to errors also.

Reply:
Also, cache_readFromMemOnMiss could be use here, I think.